File: reproduce.mjs
import { closeSync, constants, fstatSync, openSync, readFileSync, readSync, writeSync } from 'node:fs';
import { performance } from 'node:perf_hooks';

const [path, count = '10000'] = process.argv.slice(2);
const writes = Number(count);
if (!path || process.argv.length > 4 || !Number.isSafeInteger(writes) || writes < 1) {
  console.error('Usage: node reproduce.mjs FILE [WRITES=10000]');
  process.exit(1);
}

// Fixed input offsets make the workload identical across runtimes and runs.
const offsets = JSON.parse(readFileSync(new URL('./offsets.json', import.meta.url), 'utf8'));
if (writes > offsets.length) throw new Error(`Maximum write count: ${offsets.length}`);

const payload = Buffer.from('Heloo\n');
let fd = openSync(path, 'r');
try {
  const { size } = fstatSync(fd);
  if (size !== 100 * 1048576) throw new Error('Expected the included 100 MiB test file');
  const buffer = Buffer.alloc(1024 * 1024);
  const readStart = performance.now();
  while (readSync(fd, buffer, 0, buffer.length, null) > 0) {}
  console.log(`File: ${size / 1048576} MiB; full pre-read: ${(performance.now() - readStart).toFixed(2)} ms`);
  closeSync(fd);
  fd = undefined;
  fd = openSync(path, constants.O_WRONLY);

  const slow = [];
  let longest = 0;
  const start = performance.now();
  for (let n = 1; n <= writes; n++) {
    const offset = offsets[n - 1];
    const before = performance.now();
    const written = writeSync(fd, payload, 0, payload.length, offset);
    const elapsed = performance.now() - before;
    if (written !== payload.length) throw new Error(`Short write: ${written}`);
    longest = Math.max(longest, elapsed);
    if (elapsed > 100) slow.push({ write: n, offset, milliseconds: elapsed });
  }
  const total = performance.now() - start;
  console.log(`${writes.toLocaleString('en-US')} random 6-byte writes: ${total.toFixed(2)} ms`);
  console.log(`Logical bytes written: ${writes * payload.length}`);
  console.log(`Longest write: ${longest.toFixed(2)} ms`);
  console.log(`Writes exceeding 100 ms: ${slow.length}`);
  for (const { write, offset, milliseconds } of slow) {
    console.log(`  Write #${write}, offset ${offset}: ${milliseconds.toFixed(2)} ms`);
  }
} finally {
  if (fd !== undefined) closeSync(fd);
}
console.log('No explicit fsync; close and subsequent unmount flushing are outside write timing.');

File: reproduce.py
import argparse
import os
import random
import time

parser = argparse.ArgumentParser(description="Reproduce small random-write stalls on Archil")
parser.add_argument("file", help="Path to the included 100 MiB file copied onto Archil")
parser.add_argument("writes", type=int, nargs="?", default=4193)
args = parser.parse_args()
if args.writes < 1:
    parser.error("writes must be positive")

size = os.stat(args.file).st_size
if size != 100 * 1048576:
    parser.error("expected the included 100 MiB test file")

start = time.monotonic()
with open(args.file, "rb", buffering=0) as f:
    while f.read(1048576):
        pass
print(f"File: 100 MiB; full pre-read: {(time.monotonic() - start) * 1000:.2f} ms")

rng = random.Random(16001)
payload = b"Heloo\n"
slow = []
longest = 0
fd = os.open(args.file, os.O_WRONLY)
try:
    start = time.monotonic()
    for n in range(1, args.writes + 1):
        offset = rng.randrange(size - len(payload) + 1)
        before = time.monotonic()
        written = os.pwrite(fd, payload, offset)
        elapsed = time.monotonic() - before
        if written != len(payload):
            raise RuntimeError(f"Short write: {written}")
        longest = max(longest, elapsed)
        if elapsed > 0.1:
            slow.append((n, offset, elapsed))
    total = time.monotonic() - start
finally:
    os.close(fd)

print(f"{args.writes:,} random 6-byte writes (seed 16001): {total * 1000:.2f} ms")
print(f"Logical bytes written: {args.writes * len(payload)}")
print(f"Longest write: {longest * 1000:.2f} ms")
print(f"Writes exceeding 100 ms: {len(slow)}")
for n, offset, elapsed in slow:
    print(f"  Write #{n}, offset {offset}: {elapsed * 1000:.2f} ms")
print("No explicit fsync; close and subsequent unmount flushing are outside write timing.")
